### PR TITLE
BUG: is_dtype_equal(dtype, "string[pyarrow]") raises if pyarrow not installed

### LINKED ISSUE #44276: BUG: is_dtype_equal(dtype, "string[pyarrow]") raises if pyarrow not installed
### 

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [X] I have confirmed this bug exists on the master branch of pandas.

### Reproducible Example

```python
https://dev.azure.com/pandas-dev/pandas/_build/results?buildId=68369&view=logs&j=915dbb23-6529-54ae-059d-ec0b59ba10e8&t=86b091c2-17fe-52fe-2454-5f56a371e155&l=2889
```

### Issue Description

Looks like it tries to construct the dtype and raises an ImportError

### Expected Behavior

False

### Installed Versions

<details>

Replace this line with the output of pd.show_versions()

</details>

### PR DESCRIPTION
**Pull request by shubham11941140**
- [x] closes #44276
- [x] tests added / passed
- [x] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit) for how to run them
- [x] whatsnew entry

Added instance to check string preventing import error

### CODE REVIEW

File: pandas/tests/base/test_misc.py [hunk 1]
```diff
@@ -150,6 +151,8 @@ def test_access_by_position(index):
     assert index[-1] == index[size - 1]
 
     msg = f"index {size} is out of bounds for axis 0 with size {size}"
+
+    assert not is_dtype_equal(index.dtype, "string[pyarrow]")
```
Comment (jreback) [inline pandas/tests/base/test_misc.py:155]: can you move to the tests for is_dtype_equal,

File: pandas/core/dtypes/common.py [hunk 2]
```diff
@@ -601,7 +601,8 @@ def is_dtype_equal(source, target) -> bool:
             #  is_dtype_equal(CDT, "category") and CDT == "category"
             try:
                 src = get_dtype(source)
-                if isinstance(src, ExtensionDtype):
+                if isinstance(src, ExtensionDtype) or isinstance(src, np.dtype):
```
Comment (jreback) [inline pandas/core/dtypes/common.py:604]: can you add the example to the doc-string

**Review by jreback — CHANGES_REQUESTED** [review]
pls add a whatsnew note

**Comment by jreback** [comment]
do we actually have a test which exercises this? e.g. not on a build where pyarrow is installed?

**Comment by shubham11941140** [comment]
> do we actually have a test which exercises this? e.g. not on a build where pyarrow is installed?

I do not have pyarrow running on my system and got the initial import error. Actually I am able to run this on a system where pyarrow is not installed. It does not matter as pyarrow is not imported, so installation will not matter.

File: pandas/tests/base/test_misc.py (hunk 1 above)
Comment (shubham11941140) [reply pandas/tests/base/test_misc.py:155]: Done

File: pandas/core/dtypes/common.py (hunk 2 above)
Comment (shubham11941140) [reply pandas/core/dtypes/common.py:606]: Done

**Comment by shubham11941140** [comment]
Whatsnew note added

File: pandas/core/dtypes/common.py (hunk 2 above)
Comment (jbrockmendel) [reply pandas/core/dtypes/common.py:606]: single isinstance call `isinstance(src, (ExtensionDtype, np.dtype))`

File: pandas/tests/dtypes/test_dtypes.py [hunk 3]
```diff
@@ -115,6 +115,10 @@ def test_construction_from_string(self, dtype):
         with pytest.raises(TypeError, match=msg):
             CategoricalDtype.construct_from_string("foo")
 
+    def test_pyarrow_string_import_error(self, dtype):
+        # GH-44276
```
Comment (jbrockmendel) [inline pandas/tests/dtypes/test_dtypes.py:119]: can you flesh this out a tiny bit e.g. 'dont raise if pyarrow is not installed'

Comment (shubham11941140) [reply pandas/tests/dtypes/test_dtypes.py:119]: Done

File: pandas/core/dtypes/common.py (hunk 2 above)
Comment (shubham11941140) [reply pandas/core/dtypes/common.py]: Done

File: doc/source/whatsnew/v1.3.5.rst [hunk 4]
```diff
@@ -24,7 +24,7 @@ Fixed regressions
 
 Bug fixes
 ~~~~~~~~~
--
+- Fixed bug in :bug:`is_dtype_equal` raises if pyarrow not installed (:issue:`44327`)
```
Comment (jbrockmendel) [inline doc/source/whatsnew/v1.3.5.rst:27]: raises -> raising

Comment (shubham11941140) [reply doc/source/whatsnew/v1.3.5.rst]: Done

File: pandas/tests/dtypes/test_dtypes.py [hunk 5]
```diff
@@ -115,6 +115,12 @@ def test_construction_from_string(self, dtype):
...
+        assert not is_dtype_equal(dtype, "string[pyarrow]")
+        with pytest.raises(ImportError, match="pyarrow"):
```
Comment (jreback) [inline pandas/tests/dtypes/test_dtypes.py:121]: umm we don't do this, rather use the decorator to skip the test

**Review by jreback — CHANGES_REQUESTED** [review]

File: doc/source/whatsnew/v1.3.5.rst [hunk 6]
```diff
@@ -24,7 +24,7 @@ Fixed regressions
 
 Bug fixes
 ~~~~~~~~~
--
+- Fixed bug in :bug:`is_dtype_equal` raising if pyarrow not installed (:issue:`44327`)
```
Comment (jreback) [inline doc/source/whatsnew/v1.3.5.rst:27]: this goes in 1.4

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by jreback** [comment]
pls merge master as well

File: pandas/tests/dtypes/test_dtypes.py [hunk 7]
```diff
@@ -115,6 +115,12 @@ def test_construction_from_string(self, dtype):
         with pytest.raises(TypeError, match=msg):
             CategoricalDtype.construct_from_string("foo")
 
+    def test_pyarrow_string_import_error(self, dtype):
```
Comment (jorisvandenbossche) [inline pandas/tests/dtypes/test_dtypes.py:118]: Another note: this test is added to `TestCategoricalDtype`, so categorical specific tests. I think we should move it to a more generic location. For example in `tests/dtypes/test_common.py` there is a `test_dtype_equal` that tests the basic functionality of `is_dtype_equal`. Maybe add the new test there (or in the arrow string dtype specific tests, @jbrockmendel any preference?)

File: pandas/tests/dtypes/test_dtypes.py (hunk 5 above)
Comment (shubham11941140) [reply pandas/tests/dtypes/test_dtypes.py]: Done

File: doc/source/whatsnew/v1.3.5.rst (hunk 6 above)
Comment (shubham11941140) [reply doc/source/whatsnew/v1.3.5.rst]: Done

File: pandas/tests/dtypes/test_dtypes.py (hunk 7 above)
Comment (shubham11941140) [reply pandas/tests/dtypes/test_dtypes.py:118]: Request you to inform me where do I exactly put the test, in which file and which function

File: pandas/tests/dtypes/test_dtypes.py [hunk 8]
```diff
@@ -115,6 +115,11 @@ def test_construction_from_string(self, dtype):
...
+        # GH-44276
+        assert not is_dtype_equal(dtype, "string[pyarrow]")
+        pytest.importorskip("pyarrow")
```
Comment (jreback) [inline pandas/tests/dtypes/test_dtypes.py:121]: use the decorator on the function itself to skip

Comment (shubham11941140) [reply pandas/tests/dtypes/test_dtypes.py]: Done

File: pandas/tests/dtypes/test_dtypes.py [hunk 9]
```diff
@@ -115,6 +115,11 @@ def test_construction_from_string(self, dtype):
         with pytest.raises(TypeError, match=msg):
             CategoricalDtype.construct_from_string("foo")
 
+    @pytest.importorskip("pyarrow")
```
Comment (jorisvandenbossche) [inline pandas/tests/dtypes/test_dtypes.py:118]: I don't think this test should be skipped if pyarrow is not installed? Because I understood the issue that checking the dtype equality currently raises if pyarrow is *not* installed, while it should return False instead. 

So both if pyarrow is installed or not, this test should pass and `is_dtype_equal` should return False. 

I think you can simply remove the decorator.

File: pandas/tests/dtypes/test_dtypes.py (hunk 7 above)
Comment (jorisvandenbossche) [reply pandas/tests/dtypes/test_dtypes.py:119]: You can put it in `tests/dtypes/test_common.py`, just after the existing `test_dtype_equal` test.

**Review by jorisvandenbossche — CHANGES_REQUESTED** [review]

File: pandas/tests/dtypes/test_dtypes.py (hunk 9 above)
Comment (shubham11941140) [reply pandas/tests/dtypes/test_dtypes.py]: Done

**Comment by shubham11941140** [comment]
Done, put the test into a new file as specified

File: pandas/tests/dtypes/test_dtypes.py (hunk 7 above)
Comment (shubham11941140) [reply pandas/tests/dtypes/test_dtypes.py]: Done

File: pandas/tests/dtypes/test_common.py [hunk 10]
```diff
@@ -128,6 +128,12 @@ def test_dtype_equal(name1, dtype1, name2, dtype2):
         assert not com.is_dtype_equal(dtype1, dtype2)
 
 
+@pytest.mark.parametrize("name,dtype", list(dtypes.items()), ids=lambda x: str(x))
```
Comment (jorisvandenbossche) [inline pandas/tests/dtypes/test_common.py:131]: To the above `dtypes` dict, can you add an entry for `"string": pd.StringDtype()` ?

Comment (shubham11941140) [reply pandas/tests/dtypes/test_common.py:131]: This will be a mistake because isinstance(pd.StringDtype(), str) is False. Adding such an entry in taking to the wrong location in the code segment.

Comment (jorisvandenbossche) [reply pandas/tests/dtypes/test_common.py:131]: >  isinstance(pd.StringDtype(), str)

Where is this being checked?

My question is to ensure that `is_dtype_equal(pd.StringDtype(), "string[pyarrow]")` correctly works, since that was the original bug report.

Comment (shubham11941140) [reply pandas/tests/dtypes/test_common.py:131]: if isinstance(target, str):
        if not isinstance(source, str):
            # GH#38516 ensure we get the same behavior from
            #  is_dtype_equal(CDT, "category") and CDT == "category"
            # GH-44276 ensures behaviour is consistent
            # is_dtype_equal (dtype, "string[pyarrow]")
            try:
                src = get_dtype(source)
                if isinstance(src, ExtensionDtype):
                if isinstance(src, (ExtensionDtype, np.dtype)):


In this code  ```if not isinstance(source, str):``` returns True when source is ```pd.StringDtype()```

Comment (jorisvandenbossche) [reply pandas/tests/dtypes/test_common.py:131]: This line returns True for _all_ cases in the current test. Also the other `dtype` values in the parametrization of this test are actual dtype objects, not strings.

Comment (shubham11941140) [reply pandas/tests/dtypes/test_common.py:131]: Agreed, but on adding this line into the parameters, this ```if not isinstance(source, str):``` returns True which is wrong. Shouldn't ```isinstance(source, str)``` be True if source is ```pd.StringDtype()```

Comment (jorisvandenbossche) [reply pandas/tests/dtypes/test_common.py:131]: No, `isinstance(source, str)` checks whether `source` is an actual python string object, while `pd.StringDtype` is a dtype object, which is a different class, and thus this is expected to return False (the fact that StringDtype is a dtype object for an array that stores strings doesn't matter in that case)

Comment (shubham11941140) [reply pandas/tests/dtypes/test_common.py:131]: > Done

Comment (shubham11941140) [reply pandas/tests/dtypes/test_common.py:131]: I added the testcase and fixed the Import Bug as well

**Comment by shubham11941140** [comment]
@jorisvandenbossche @jreback any updates for me on this?

File: doc/source/whatsnew/v1.4.0.rst [hunk 11]
```diff
@@ -509,7 +509,7 @@ Conversion
 
 Strings
 ^^^^^^^
--
+- Fixed bug in :bug:`is_dtype_equal` raising if pyarrow not installed (:issue:`44327`)
```
Comment (jreback) [inline doc/source/whatsnew/v1.4.0.rst:512]: this won't render & is not public facing anyhow. remove the note

File: pandas/core/dtypes/common.py [hunk 12]
```diff
@@ -596,12 +596,17 @@ def is_dtype_equal(source, target) -> bool:
     False
     """
     if isinstance(target, str):
-        if not isinstance(source, str):
+        if get_dtype(source) == "string":
```
Comment (jreback) [inline pandas/core/dtypes/common.py:599]: why is this needed?

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/dtypes/common.py (hunk 12 above)
Comment (shubham11941140) [reply pandas/core/dtypes/common.py:599]: It is directly False, there is no need for further checking for this.

File: doc/source/whatsnew/v1.4.0.rst (hunk 11 above)
Comment (shubham11941140) [reply doc/source/whatsnew/v1.4.0.rst:512]: I did not understand the part where "is not public facing anyhow.". I will anyhow remove it.

File: pandas/core/dtypes/common.py (hunk 12 above)
Comment (jreback) [reply pandas/core/dtypes/common.py:599]: the reason i ask is this might be a performance issue here (this code is hit a lot so we care about the code path)

Comment (shubham11941140) [reply pandas/core/dtypes/common.py]: I have changed the code to remove this code segment, now there should not be a problem.

File: doc/source/whatsnew/v1.4.0.rst (hunk 11 above)
Comment (shubham11941140) [reply doc/source/whatsnew/v1.4.0.rst:512]: Do I remove this?

Comment (jorisvandenbossche) [reply doc/source/whatsnew/v1.4.0.rst:512]: I think you can keep it, but reword it a bit to not mention `is_dtype_equal` explicitly. Something like "Fixed a bug in checking if a dtype is the `string[pyarrow]` dtype incorrectly raising an error when pyarrow is not installed (:issue:...)"

File: pandas/core/dtypes/common.py [hunk 13]
```diff
@@ -599,10 +599,13 @@ def is_dtype_equal(source, target) -> bool:
...
             #  is_dtype_equal(CDT, "category") and CDT == "category"
+            # GH-44276 ensures behaviour is consistent
+            # is_dtype_equal (dtype, "string[pyarrow]")
             try:
                 src = get_dtype(source)
-                if isinstance(src, ExtensionDtype):
-                    return src == target
+                if isinstance(src, (ExtensionDtype, np.dtype)):
+
+                    return type(src) == type(target)
```
Comment (jorisvandenbossche) [inline pandas/core/dtypes/common.py:608]: You can't change this to only checking the class of the dtype, we want to check the equality of the instances. See the failing tests

File: pandas/core/dtypes/common.py [hunk 14]
```diff
@@ -599,10 +599,13 @@ def is_dtype_equal(source, target) -> bool:
...
             except (TypeError, AttributeError):
```
Comment (jorisvandenbossche) [inline pandas/core/dtypes/common.py:609]: I think you can probably catch the ImportError here (so it always returns False if the above raises an ImportError)

Comment (shubham11941140) [reply pandas/core/dtypes/common.py]: Done

File: doc/source/whatsnew/v1.4.0.rst (hunk 11 above)
Comment (shubham11941140) [reply doc/source/whatsnew/v1.4.0.rst]: Done

File: pandas/core/dtypes/common.py (hunk 13 above)
Comment (shubham11941140) [reply pandas/core/dtypes/common.py]: Reverted Back to the Original Code, saw the mistake.

File: pandas/core/dtypes/common.py [hunk 15]
```diff
@@ -599,11 +599,14 @@ def is_dtype_equal(source, target) -> bool:
         if not isinstance(source, str):
             # GH#38516 ensure we get the same behavior from
             #  is_dtype_equal(CDT, "category") and CDT == "category"
+            # GH-44276 ensures behaviour is consistent
+            # is_dtype_equal (dtype, "string[pyarrow]")
             try:
                 src = get_dtype(source)
-                if isinstance(src, ExtensionDtype):
+                if isinstance(src, (ExtensionDtype, np.dtype)):
```
Comment (jorisvandenbossche) [inline pandas/core/dtypes/common.py:606]: is this change still needed?

Comment (shubham11941140) [reply pandas/core/dtypes/common.py:606]: It is needed because dtypes such as ```int64``` ```float64``` and other dtypes actually are ```np.dtype```, so if I am not adding this check, the test is failing.

**Comment by shubham11941140** [comment]
@jorisvandenbossche @jreback any updates for me on this?

**Comment by jreback** [comment]
@shubham11941140 you need to merge master

**Comment by shubham11941140** [comment]
Done.

File: pandas/core/dtypes/common.py (hunk 15 above)
Comment (jorisvandenbossche) [reply pandas/core/dtypes/common.py]: > It is needed because dtypes such as int64 float64 and other dtypes actually are np.dtype, so if I am not adding this check, the test is failing.

It is not needed if we catch ImportError more generally, I think; I pushed some small commits to verify this.

**Review by jorisvandenbossche — APPROVED** [review]

**Comment by jorisvandenbossche** [comment]
Thanks @shubham11941140
